src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexMatcher.cs

```diff
@@ -343,6 +380,23 @@ private DfaMatchingState<TSetType> CreateNewTransition(DfaMatchingState<TSetType
             }
         }
 
+        private List<(DfaMatchingState<TSetType>, List<DerivativeEffect>)> CreateNewCapturingTransitions(DfaMatchingState<TSetType> state, TSetType minterm, int offset)
+        {
+            Debug.Assert(_builder._capturingDelta is not null);
+            lock (this)
+            {
+                // check if meanwhile delta[offset] has become defined possibly by another thread
+                List<(DfaMatchingState<TSetType>, List<DerivativeEffect>)> p = _builder._capturingDelta[offset];
+                if (p is null)
+                {
+                    // this is the only place in code where the Next method is called in the matcher
+                    _builder._capturingDelta[offset] = p = new List<(DfaMatchingState<TSetType>, List<DerivativeEffect>)>(state.AntimirovEagerNextWithEffects(minterm));
```

Comment:
Technically this should be:
```suggestion
                    p = new List<(DfaMatchingState<TSetType>, List<DerivativeEffect>)>(state.AntimirovEagerNextWithEffects(minterm));
                    Volatile.Write(ref _builder._capturingDelta[offset], p);
```
but in practice the runtime provides a stronger memory model here than ECMA guarantees.  I think it's still worth doing the volatile thing.